ibis/backends/__init__.py

```diff
@@ -473,6 +473,46 @@ def to_parquet(
                 for batch in batch_reader:
                     writer.write_batch(batch)
 
+    @util.experimental
+    def to_parquet_dir(
+        self,
+        expr: ir.Table,
+        path: str | Path,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Write the results of executing the given expression to a parquet file in a directory.
+
+        This method is eager and will execute the associated expression
+        immediately.
+
+        Parameters
+        ----------
+        expr
+            The ibis expression to execute and persist to parquet.
+        path
+            The data source. A string or Path to the directory where the parquet file will be written.
+        params
+            Mapping of scalar parameter expressions to value.
+        **kwargs
+            Additional keyword arguments passed to pyarrow.parquet.ParquetWriter
+
+        https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html
+
+        """
+        self._import_pyarrow()
+        import pyarrow.parquet as pq
+
+        dir_path = Path(path)
+        dir_path.mkdir(parents=True, exist_ok=True)
+
+        with expr.to_pyarrow_batches(params=params) as batch_reader:
+            for i, batch in enumerate(batch_reader):
+                file_path = f"{dir_path}/out_{i}.parquet"
```

Comment:
Naming files this way will lead to annoying consequences like sorting file names in lexicographic order instead of the sequential order that they are written.

Since we don't know how many batches there are, you'll have to write them out and then rename them using a format string padded with zeroes that reflect the order of magnitude of the number of files written.

Reply:
If we use pyarrow then users can deal with this annoyance themselves using the `basename_template` parameter.